### PR TITLE
refactor(encoding): split match_generator monolith + separate tests from production

### LINKED ISSUE #441: refactor(encoding): split match_generator.rs monolith into per-concern modules
## Summary

Split the `encoding/match_generator.rs` monolith into per-concern modules so no
file in `encoding/` exceeds ~1500 LoC and `match_generator.rs` is left as
orchestration only. This is the structural goal carried (and unmet) by the
encoder-rewrite umbrella; tracked standalone now that the umbrella's shipped
phases are closed.

## Current state

`match_generator.rs` is **13,020 LoC** (it has *grown* from ~9,000 since the
rewrite began — the Phase 7 perf work keeps adding matcher/optimizer code to the
monolith). Nine `encoding/` files exceed 1500 LoC:

```
13020  match_generator.rs
 4859  frame_compressor.rs
 3369  simple/fast_matcher.rs
 3278  dfast/mod.rs
 2918  blocks/compressed.rs
 2719  match_table/storage.rs
 2688  simple/fast_kernel/kernel.rs
 2590  row/mod.rs
 1870  streaming_encoder.rs
```

## Internal structure of match_generator.rs (extractable concerns)

- **158-1263 (~1100 LoC)** — level config: `HcConfig` / `DfastConfig` /
  `FastConfig` / `LevelParams` + the `LEVEL_TABLE`. → `encoding/levels/config.rs`.
- **1319-2040** — `MatchGeneratorDriver` reset + dict prime / capture / restore.
  → `encoding/dict_prime.rs`.
- **3600-8084 (~4500 LoC)** — `HcMatchGenerator` + its impl (HC chain / BT walk
  glue over the already-split `MatchTable` / `HcMatcher` / `BtMatcher`).
  → `encoding/hc/`.
- **8085-13020 (~4900 LoC)** — `MatchGeneratorDriver` optimal-parse tail
  (`build_optimal_plan` DP, cost model, price tables). → `encoding/opt/`.
- Residual driver orchestration + `Matcher` impl stays in `match_generator.rs`
  (target ≤ 1500 LoC).

## Approach

Pure mechanical extraction, **one concern per PR**, no behaviour change:

1. Extract level config → `levels/config.rs`.
2. Extract optimal-parse DP → `opt/`.
3. Extract HcMatchGenerator → `hc/`.
4. Extract dict prime/capture/restore → `dict_prime.rs`.
5. Sweep the other > 1500 LoC files (frame_compressor, dfast, compressed, row)
   in follow-up PRs as each is touched.

## Acceptance criteria

- [ ] `match_generator.rs` ≤ 1500 LoC (orchestration only)
- [ ] No `encoding/` file exceeds 1500 LoC (or a tracked follow-up exists per file)
- [ ] **Byte-identical compressed output** on the corpus before/after every PR
- [ ] `cargo nextest run --features dict_builder` 100% pass on every commit
- [ ] clippy clean on x86_64 + aarch64 + i686

## Constraints

- Mechanical move only — no logic change, no perf change. The byte-identity gate
  is the safety net for each extraction.
- Do not block the active perf work (#434 / #323 / #178 / #337); coordinate so a
  split PR rebases cleanly over in-flight matcher changes (extract the *settled*
  concerns first — config table, then opt DP).

## Context

Spun out of the encoder-rewrite umbrella after its Phases 1-6 (structural split,
const-generic `Strategy`, LDM, block-splitting) shipped. The remaining perf lanes
are tracked by #434, #323, #178, #337, #205; this issue tracks only the
file-size / module-boundary cleanup.

## Time estimate

`3d` (careful mechanical extraction of ~10k LoC across 4 modules, byte-identity
re-verification per step, + clippy/triple-arch CI per PR).

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Two related refactors, no production logic changes (compress/decompress output is byte-identical; the suite runs the same 824 cases throughout).

### 1. Split the `match_generator.rs` monolith (closes #441)

The ~13k-line driver is broken into an orchestrator plus per-concern modules:

| Module | Concern |
|--------|---------|
| `match_generator/mod.rs` | driver orchestrator |
| `encoding/levels/config.rs` | level table, resolvers, estimators |
| `encoding/hc/priceset.rs` | per-CPU price-set SIMD kernels |
| `encoding/hc/generator.rs` | `HcMatchGenerator` + chain matcher |
| `encoding/hc/optimal.rs` | optimal-parse DP + body macros |
| `match_generator/dict_prime.rs` | dictionary prime / CDict-equivalent snapshot lifecycle |

`dict_prime` is a child module so it reaches the driver's private, invariant-bearing snapshot state (`primed` / `reset_shape` / `PrimedKey`) directly, with zero field-visibility widening; the `Matcher` trait surface forwards to it via thin `#[inline]` delegations.

### 2. Separate test code from production across the crate

Every inline test module and module-scope `#[test]` is relocated into a sibling test file, leaving production sources with code only plus a `#[cfg(test)] mod ...;` declaration. Covers ~75 files across `zstd` and `cli`. Test-support items that are methods/fields on production types (test-only accessors over private state) intentionally stay with their type.

## Testing

- `cargo nextest run -p structured-zstd --features dict_builder` — 824 passed
- `cargo nextest run -p structured-zstd-cli` — 22 passed
- `cargo clippy --tests` + `cargo fmt` clean
- `cargo check --workspace --tests` clean
- `x86_64-apple-darwin` cross-checked for the arch-gated tests (row tag-mask, reverse-bit-reader BMI2 helper) that the local aarch64 build compiles out

Closes #441

## Summary by CodeRabbit

* **Tests**
  * Expanded and reorganized unit test coverage across the CLI and core library, including formatting helpers, argument parsing behavior, and extensive decoding/encoding/dictionary edge cases.
  * Added new targeted regression tests to validate byte-accurate operations, error messages, and boundary conditions.

* **Refactor**
  * Moved large in-file test implementations into dedicated test modules to improve maintainability and keep production logic unchanged.

### CODE REVIEW

File: zstd/src/decoding/frame_decoder/tests.rs [hunk 1]
```diff
@@ -0,0 +1,2362 @@
...
+    // rejected before any state is restored.
+    let (compressed, full, info) = multi_block_fixture();
+    let nblocks = info.blocks.len() as u32;
+    let n = (nblocks / 2).max(1);
+    let st = emit_resume_state_at(&compressed, n); // active_dictionary_id = None
+    let output_offset = info.decompressed_byte_range(n as usize).unwrap().start as usize;
+
+    let raw = std::fs::read("./dict_tests/dictionary").expect("dictionary fixture");
+    let dict = crate::decoding::dictionary::Dictionary::decode_dict(&raw).expect("parse dict");
+    let dict_id = dict.id;
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/decoding/frame_decoder/tests.rs]: Fixed in this round: replaced the CWD-relative `std::fs::read("./dict_tests/dictionary")` with `include_bytes!("../../../dict_tests/dictionary")`, matching the compile-time fixtures the rest of the file uses, so the test no longer depends on the harness working directory.

File: cli/src/tests.rs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        bench_end: 0,
+        bench_secs: 1.0,
+        long: false,
+    };
+    assert_eq!(
+        derive_output_path(&opts, Path::new("archive.tar.zst")).unwrap(),
+        PathBuf::from("archive.tar")
+    );
+    assert!(derive_output_path(&opts, Path::new("noext")).is_err());
+}
```
[the remark this answers is not in the document]
Comment (polaz) [reply cli/src/tests.rs]: Fixed: ran `cargo fmt --all` (the per-package run had missed the extracted test files). The leading blank line each file kept from its inline `mod` origin is removed and `cargo fmt --all -- --check` is clean.
